src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,485 @@
+            return invalid_entry_exception(); // Only empty directories may be deleted
+        }
+
+        _metadata_log.memory_only_delete_inode(entry.inode);
+        return now();
+    }
+
+    future<> bootstrap_small_write() {
+        ondisk_small_write_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.inode)) {
+            return invalid_entry_exception();
+        }
+
+        auto data_opt = _curr_checkpoint.read_tmp_buff(entry.length);
+        if (not data_opt) {
+            return invalid_entry_exception();
+        }
+        temporary_buffer<uint8_t>& data = *data_opt;
+
+        if (not std::holds_alternative<inode_info::file>(_metadata_log._inodes[entry.inode].contents)) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_small_write(entry.inode, entry.offset, std::move(data));
+        _metadata_log.memory_only_update_mtime(entry.inode, entry.mtime_ns);
+        return now();
+    }
+
+    future<> bootstrap_medium_write() {
+        // TODO: will be very similar to SMALL_WRITE
+        assert(false && "Not implemented");
+        return now();
+    }
+
+    future<> bootstrap_large_write() {
+        // TODO: will be very similar to SMALL_WRITE
+        assert(false && "Not implemented");
+        return now();
+    }
+
+    future<> bootstrap_large_write_without_mtime() {
+        // TODO: will be very similar to SMALL_WRITE
+        assert(false && "Not implemented");
+        return now();
+    }
+
+    future<> bootstrap_truncate() {
+        ondisk_truncate entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.inode)) {
+            return invalid_entry_exception();
+        }
+
+        if (not std::holds_alternative<inode_info::file>(_metadata_log._inodes[entry.inode].contents)) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_truncate(entry.inode, entry.size);
+        _metadata_log.memory_only_update_mtime(entry.inode, entry.mtime_ns);
+        return now();
+    }
+
+    future<> bootstrap_mtime_update() {
+        ondisk_mtime_update entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.inode)) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_update_mtime(entry.inode, entry.mtime_ns);
+        return now();
+    }
+
+    future<> bootstrap_add_dir_entry() {
+        ondisk_add_dir_entry_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.dir_inode) or not inode_exists(entry.entry_inode)) {
+            return invalid_entry_exception();
+        }
+
+        sstring dir_entry_name;
+        if (not _curr_checkpoint.read_sstring(dir_entry_name, entry.entry_name_length)) {
+            return invalid_entry_exception();
+        }
+
+        // Only files may be linked as not to create cycles
```

Comment:
We cannot create a directory first and then add it to another directory. Maybe instead of completely forbidding directories we could forbid adding directory that is already linked somewhere?

Reply:
So far only create_inode_as_dir_entry is allowed to be used to create directories. I don't think we want to allow / need linking directories.(add_dir_entry will be used for linking files, create_inode_as_dir_entry is used to create and link simultaneously.

Reply:
So I would at least add a comment about that decision because for now, it is a little confusing I think.

Reply:
Hardlinks for directories are almost never implemented in file systems (never seen one), since they only create trouble, so I'm fine with leaving the code as is - the comment above that it's forbidden because it could create cycles is good enough.